packages/util/src/bytes.ts

```diff
@@ -122,48 +123,72 @@ export const bigIntToBytes = (num: bigint, littleEndian = false): Uint8Array =>
 
 /**
  * Pads a `Uint8Array` with zeros till it has `length` bytes.
- * Truncates the beginning or end of input if its length exceeds `length`.
+ * Throws if input length exceeds target length, unless allowTruncate is true.
  * @param {Uint8Array} msg the value to pad
  * @param {number} length the number of bytes the output should be
- * @param {boolean} right whether to start padding form the left or right
+ * @param {boolean} right whether to start padding from the left or right
+ * @param {boolean} allowTruncate whether to allow truncation if msg exceeds length
  * @return {Uint8Array}
  */
-const setLength = (msg: Uint8Array, length: number, right: boolean): Uint8Array => {
-  if (right) {
-    if (msg.length < length) {
-      return new Uint8Array([...msg, ...new Uint8Array(length - msg.length)])
-    }
-    return msg.subarray(0, length)
-  } else {
-    if (msg.length < length) {
-      return new Uint8Array([...new Uint8Array(length - msg.length), ...msg])
+const setLength = (
+  msg: Uint8Array,
+  length: number,
+  right: boolean,
+  allowTruncate: boolean,
+): Uint8Array => {
+  if (msg.length > length) {
+    if (!allowTruncate) {
+      throw EthereumJSErrorWithoutCode(
+        `Input length ${msg.length} exceeds target length ${length}. Use allowTruncate option to truncate.`,
+      )
     }
-    return msg.subarray(-length)
+    return right ? msg.subarray(0, length) : msg.subarray(-length)
   }
+  if (msg.length < length) {
+    return right
+      ? new Uint8Array([...msg, ...new Uint8Array(length - msg.length)])
+      : new Uint8Array([...new Uint8Array(length - msg.length), ...msg])
+  }
+  return msg
```

Comment:
I had to check here if the original returned a copy, but `subarray` returns a view of the array so this is not the case (so this is fine 😄 👍 )